[crossgen2] Fix memory leak in _corInfoImpls.

t-mustafin opened the pull request:
This pull request fixes crossgen2 potentially memory leak.

`ConditionalWeakTable  _corInfoImpls` elements keeps `_compilation` reference which keeps reference to whole table `_corInfoImpls`. The circular reference together with issue #12255 potentionally leads to memory leak in crossgen2 if that table is created more than once. As example `_corInfoImpls` table is created several times in pipeline mode #37411. `WeakReference` allows GC to collect the table and prevents the leak.

cc @jkotas @davidwrighton @MichalStrehovsky @gbalykov @alpencolt

jkotas commented:
I think it would be better to have a `Dispose` method on Compilation to break these cycles and allow things to be garbage collected.

t-mustafin committed:
[crossgen2] Implement Dispose in Compilation.

_corInfoImpls elements keeps _compilation reference which keeps reference to whole table _corInfoImpls. The circular referene together with issue #12255 potentionally leads to memory leak in crossgen2 if that table is created more than once. Dispose() method clears the table and prevents memory leak.

Signed-off-by: Timur Mustafin <t.mustafin@partner.samsung.com>

jkotas commented on src/coreclr/tools/aot/crossgen2/Program.cs:
```diff
@@ -632,6 +632,7 @@ private int Run(string[] args)
 
                 if (_commandLineOptions.DgmlLogFileName != null)
                     compilation.WriteDependencyLog(_commandLineOptions.DgmlLogFileName);
+                compilation.Dispose();
```
```suggestion

                compilation.Dispose();
```
Nit

jkotas committed:
Update src/coreclr/tools/aot/crossgen2/Program.cs

jkotas reviewed (approved):
LGTM. Thanks
